tests/Models/UserTest.php

```diff
@@ -65,24 +173,64 @@ public function testUsernameAvailableAtForNonDefaultGroup()
         $this->assertGreaterThanOrEqual($allowedAt, $user->getUsernameAvailableAt());
     }
 
-    public function testResetSessions(): void
+    /**
+     * @dataProvider dataProviderForUsernameChangeCost
+     */
+    public function testUsernameChangeCost(int $changes, int $cost)
     {
-        $user = User::factory()->create();
+        $user = User::factory()
+            ->has(UsernameChangeHistory::factory()->count($changes))
+            ->create();
 
-        // create session
-        $this->post(route('login'), ['username' => $user->username, 'password' => User::factory()::DEFAULT_PASSWORD]);
-        // sanity check
-        $this->assertNotEmpty(Store::ids($user->getKey()));
+        $this->assertSame($user->usernameChangeCost(), $cost);
+    }
 
-        // create token
-        $token = Token::factory()->create(['user_id' => $user, 'revoked' => false]);
-        $refreshToken = (new RefreshTokenFactory())->create(['access_token_id' => $token, 'revoked' => false]);
+    /**
+     * @dataProvider dataProviderForUsernameChangeCostLastChange
+     */
+    public function testUsernameChangeCostLastChange(int $years, int $cost)
+    {
+        $this->travelTo(CarbonImmutable::now()->subYears($years));
 
-        $user->resetSessions();
+        $user = User::factory()
+            ->has(UsernameChangeHistory::factory()->count(6)) // 6 = max tier + 1
+            ->create();
 
-        $this->assertEmpty(Store::ids($user->getKey()));
-        $this->assertTrue($token->fresh()->revoked);
-        $this->assertTrue($refreshToken->fresh()->revoked);
+        $this->travelBack();
+
+        $this->assertSame($user->usernameChangeCost(), $cost);
+    }
+
+    /**
+     * @dataProvider dataProviderForUsernameChangeCostType
+     */
+    public function testUsernameChangeCostType(string $type, int $cost)
+    {
+        $user = User::factory()
+            ->has(UsernameChangeHistory::factory()->state(['type' => $type]))
+            ->create();
+
+        $this->assertSame($user->usernameChangeCost(), $cost);
+    }
+
+    /**
+     * This tests the correct last UsernameChangeHistory is used when applying the cost changes.
+     *
+     * @dataProvider dataProviderForUsernameChangeCostTypeLastChange
+     */
+    public function testUsernameChangeCostTypeLastChange(string $type, int $cost)
+    {
+        $this->travelTo(CarbonImmutable::now()->subYears(1));
+
+        $user = User::factory()
+            ->has(UsernameChangeHistory::factory()->count(2))
+            ->create();
+
+        $this->travelBack();
+
+        UsernameChangeHistory::factory()->state(['type' => $type, 'user_id' => $user])->create();
+
+        $this->assertSame($user->usernameChangeCost(), $cost);
```

Comment:
the function signature is `$expected, $actual` so it should be `$cost, $user->...` here (and the rest of this file).